atc/db/worker_resource_cache.go

```diff
@@ -124,65 +124,64 @@ func (workerResourceCache WorkerResourceCache) FindByID(runner sq.Runner, id int
 	}, true, nil
 }
 
-// find should return a valid worker resource cache if its WorkerBaseResourceTypeID is not 0.
-// find should return an invalidated worker resource cache if its WorkerBaseResourceTypeID is 0 and 
-// a cache's invalid_since is later than volumeShouldBeValidBefore.
-// If there are multiple invalidated caches with invalid_since later than volumeShouldBeValidBefore, 
-// the invalidated cache with newest invalid_since should be returned.
+// find should return a valid worker resource cache if its WorkerBaseResourceTypeID is not null.
+// When volumeShouldBeValidBefore find should only return a valid worker resource cache; otherwise
+// find may return an invalidated worker resource cache whose WorkerBaseResourceTypeID is null and
+// invalid_since is later than volumeShouldBeValidBefore. If there are multiple invalidated caches
+// with invalid_since later than volumeShouldBeValidBefore, the invalidated cache with newest
+// invalid_since should be returned.
 func (workerResourceCache WorkerResourceCache) find(runner sq.Runner, volumeShouldBeValidBefore *time.Time) (*UsedWorkerResourceCache, bool, error) {
 	var id int
 	var workerBaseResourceTypeID sql.NullInt64
 	var invalidSince pq.NullTime
 
-	var idOfNewestInvalid int
-	var invalidSinceOfNewestInvalid time.Time
-
-	rows, err := psql.Select("id", "worker_base_resource_type_id", "invalid_since").
-		From("worker_resource_caches").
-		Where(sq.Eq{
-			"resource_cache_id": workerResourceCache.ResourceCache.ID(),
-			"worker_name":       workerResourceCache.WorkerName,
-		}).
+	err := psql.Select("id", "worker_base_resource_type_id", "invalid_since").
+		From("valid_caches").
+		OrderBy("priority").
+		Limit(1).
+		Prefix(
+			`WITH candidates AS (
+                     SELECT id, worker_base_resource_type_id, invalid_since
+                       FROM worker_resource_caches
+                       WHERE resource_cache_id = $1 and worker_name = $2
+                 ),
+				 valid_caches AS (
+                     SELECT id, worker_base_resource_type_id, invalid_since, 1 as priority
+                       FROM candidates
+                       WHERE worker_base_resource_type_id is not null
```

Comment:
This is great, I have been following this change in the recent past to overcome this issue in our production cluster. 

I have a quick question on this psql query. IIUC, Does this WHERE clause check for `worker_base_resource_type_id != 0`? Because, I think the `worker_base_resource_type_id` is set to `0` when a worker is either pruned or the cache is invalidated.

Reply:
@deepakmohanakrishnan07 No, `worker_base_resource_type_id` is set to NULL. You may check with your db to verify that.